File: core/src/main/java/org/ecos/reader/core/io/files/FileCollectionImpl.java
package org.ecos.reader.core.io.files;

import java.util.AbstractCollection;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;

public class FileCollectionImpl
    extends AbstractCollection<File>
    implements FileCollection
{
    private Collection<File> mCollectionOfFiles;

    FileCollectionImpl() {
        mCollectionOfFiles = new ArrayList<>();
    }

    @Override
    public boolean add(File file) {
        return mCollectionOfFiles.add(file);
    }

    @SuppressWarnings("NullableProblems")
    @Override
    public Iterator<File> iterator() {
        return mCollectionOfFiles.iterator();
    }

    @Override
    public int size() {
        return mCollectionOfFiles.size();
    }

    @Override
    public Collection<String> getPathsAsString() {
        Collection<String> result = new ArrayList<>();
        for (File item : mCollectionOfFiles) {
            result.add(item.getPath().asString());
        }
        return result;
    }
}

File: files/src/main/kotlin/org/ecos/core/files/FileCollectionImpl.kt
package org.ecos.core.files

import java.util.*

class FileCollectionImpl internal constructor() : AbstractCollection<File>(), FileCollection {
    private val mCollectionOfFiles: MutableCollection<File>

    override val pathsAsString: Collection<String>

    get() = mCollectionOfFiles.map { it.path.asString() }

    init {
        mCollectionOfFiles = ArrayList<File>()
    }

    override fun add(element: File): Boolean {
        return mCollectionOfFiles.add(element)
    }

    override val size: Int
        get() = mCollectionOfFiles.size

    override fun iterator(): MutableIterator<File> {
        return mCollectionOfFiles.iterator()
    }
}